### PR TITLE
ORC: In BuildOrcProjection field should be optional if any parent is optional

### LINKED ISSUE #961: BuildORCProjection does not support evolved schemas with added optional container types with required fields inside
If a table evolves to add a new optional container type i.e. struct, map, list with a required field inside of it, `OrcSchemaUtil.buildOrcProjection` fails with saying the the nested child field was required but not found in file. Iceberg supports such an evolution and so this should not fail.
e.g.
```java
  @Test
  public void testEvolutionNestedField() throws IOException {
    Schema baseSchema = new Schema(
        required(1, "int", Types.IntegerType.get())
    );

    Schema evolvedSchema = new Schema(
        required(1, "int", Types.IntegerType.get()),
        optional(2, "struct", Types.StructType.of(
            required(3, "long", Types.LongType.get())
        ))
    );

    ORCSchemaUtil.buildOrcProjection(evolvedSchema, ORCSchemaUtil.convert(baseSchema));
  }
```

```java
java.lang.IllegalArgumentException: Field 3 of type long is required and was not found.
	at org.apache.iceberg.orc.ORCSchemaUtil.buildOrcProjection(ORCSchemaUtil.java:305)
	at org.apache.iceberg.orc.ORCSchemaUtil.buildOrcProjection(ORCSchemaUtil.java:271)
	at org.apache.iceberg.orc.ORCSchemaUtil.buildOrcProjection(ORCSchemaUtil.java:271)
	at org.apache.iceberg.orc.ORCSchemaUtil.buildOrcProjection(ORCSchemaUtil.java:254)
	at org.apache.iceberg.orc.TestBuildOrcProjection.testEvolutionNestedField(TestBuildOrcProjection.java:152)
```

### PR DESCRIPTION
**Pull request by shardulm94**
Fixes #961

### CODE REVIEW

File: spark/benchmark/iceberg-source-orc-data-read-benchmark-result-after-vectorization.txt [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# JMH version: 1.21
```
Comment (rdblue) [inline spark/benchmark/iceberg-source-orc-data-read-benchmark-result-after-vectorization.txt:1]: Can you remove the benchmark results? I'd rather not commit them to the repo.

**Comment by rdblue** [comment]
Other than the benchmark results, this looks good to me. What do you think, @rdsr?

File: spark/benchmark/iceberg-source-orc-data-read-benchmark-result-after-vectorization.txt (hunk 1 above)
Comment (shardulm94) [reply spark/benchmark/iceberg-source-orc-data-read-benchmark-result-after-vectorization.txt:1]: Yep, that was not intentional.

**Comment by rdsr** [comment]
+1. From me as well!

**Comment by rdblue** [comment]
Merging this. Thanks, @shardulm94!

**Pull request merged**